spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -0,0 +1,531 @@
+				/*
+				 * Note: initial position setting is only supported with explicit topic assignment.
+				 * When using auto assignment (subscribe), the ConsumerRebalanceListener is not
+				 * called until we poll() the consumer.
+				 */
+				if (resetStrategy.equals(ContainerOffsetResetStrategy.EARLIEST)) {
+					this.consumer.seekToBeginning(
+							this.topicPartitions.toArray(new TopicPartition[this.topicPartitions.size()]));
+				}
+				else if (resetStrategy.equals(ContainerOffsetResetStrategy.LATEST)) {
+					this.consumer.seekToEnd(
+							this.topicPartitions.toArray(new TopicPartition[this.topicPartitions.size()]));
+				}
+				else if (resetStrategy.equals(ContainerOffsetResetStrategy.RECENT)) {
+					this.consumer.seekToEnd(
+							this.topicPartitions.toArray(new TopicPartition[this.topicPartitions.size()]));
+					for (TopicPartition topicPartition : this.topicPartitions) {
+						long newOffset = this.consumer.position(topicPartition) - recentOffset;
+						this.consumer.seek(topicPartition, newOffset);
+						if (logger.isDebugEnabled()) {
+							logger.debug("Reset " + topicPartition + " to offset " + newOffset);
+						}
+					}
+				}
+			}
+			while (running) {
+				try {
+					if (logger.isTraceEnabled()) {
+						logger.trace("Polling...");
+					}
+					ConsumerRecords<K, V> records = consumer.poll(pollTimeout);
+					if (records != null) {
+						count += records.count();
+						if (logger.isDebugEnabled()) {
+							logger.debug("Received: " + records.count() + " records");
+						}
+						Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
+						while (iterator.hasNext()) {
+							final ConsumerRecord<K, V> record = iterator.next();
+							if (this.acknowledgingMessageListener != null) {
+								this.acknowledgingMessageListener.onMessage(record, new Acknowledgment() {
+
+									@Override
+									public void acknowledge() {
+										manualAcks.add(record);
+									}
+
+								});
+							}
+							else {
+								this.listener.onMessage(record);
+							}
+							if (!this.autoCommit && ackMode.equals(AckMode.RECORD)) {
+								this.consumer.commitAsync(
+										Collections.singletonMap(new TopicPartition(record.topic(), record.partition()),
+												new OffsetAndMetadata(record.offset())), callback);
+							}
+						}
+						if (!this.autoCommit && !records.isEmpty()) {
+							if (ackMode.equals(AckMode.BATCH)) {
+								this.consumer.commitAsync(callback);
+							}
+							else if (ackMode.equals(AckMode.COUNT) && count >= ackCount) {
+ 								this.consumer.commitAsync(buildCommitMap(), callback);
+ 								count = 0;
+							}
+							else if (ackMode.equals(AckMode.TIME)
+									&& (now = System.currentTimeMillis()) - last > ackTime) {
+ 								this.consumer.commitAsync(buildCommitMap(), callback);
+								last = now;
+								this.unAcked.clear();
+							}
+							else if (!ackMode.equals(AckMode.MANUAL)) {
+								this.unAcked.add(records);
```

Comment:
On the long run, I think that we will just want a structure that keeps the highest consumed offset indexed by TopicPartition, rather than the records themselves. The ack windows can encompass hundreds of thousands if not millions of records.